src/diffwofost/physical_models/crop/leaf_dynamics.py

```diff
@@ -0,0 +1,396 @@
+        WLV  = (params.TDWI * (1-FR)) * FL
+        DWLV = torch.tensor(0., dtype=DTYPE)
+        TWLV = WLV + DWLV
+
+        # First leaf class (SLA, age and weight)
+        SLA   = torch.tensor([params.SLATB(DVS)], dtype=DTYPE)
+        LVAGE = torch.tensor([0.], dtype=DTYPE)
+        LV    = torch.stack([WLV])
+
+        # Initial values for leaf area
+        LAIEM  = LV[0] * SLA[0]
+        LASUM  = LAIEM
+        LAIEXP = LAIEM
+        LAIMAX = LAIEM
+        LAI    = LASUM + self.kiosk["SAI"] + self.kiosk["PAI"]
+
+        # Initialize StateVariables object
+        self.states = self.StateVariables(kiosk, publish=["LAI","TWLV","WLV"],
+                                          LV=LV, SLA=SLA, LVAGE=LVAGE, LAIEM=LAIEM,
+                                          LASUM=LASUM, LAIEXP=LAIEXP, LAIMAX=LAIMAX,
+                                          LAI=LAI, WLV=WLV, DWLV=DWLV, TWLV=TWLV)
+
+    def _calc_LAI(self):
+        # Total leaf area Index as sum of leaf, pod and stem area
+        SAI = self.kiosk["SAI"]
+        PAI = self.kiosk["PAI"]
+        total_LAI = self.states.LASUM + SAI + PAI
+        return total_LAI
+
+    @prepare_rates
+    def calc_rates(self, day, drv):
+        """Calculate the rates of change for the leaf dynamics."""
+        r = self.rates
+        s = self.states
+        p = self.params
+        k = self.kiosk
+
+        # Growth rate leaves
+        # weight of new leaves
+        r.GRLV = k.ADMI * k.FL
+
+        # death of leaves due to water/oxygen stress
+        r.DSLV1 = s.WLV * (1. - k.RFTRA) * p.PERDL
+
+        # death due to self shading cause by high LAI
+        DVS = self.kiosk["DVS"]
+        LAICR = 3.2 / p.KDIFTB(DVS)
+        r.DSLV2 = s.WLV * limit(0., 0.03, 0.03 * (s.LAI - LAICR) / LAICR)
+
+        # Death of leaves due to frost damage as determined by
+        # Reduction Factor Frost "RF_FROST"
+        if "RF_FROST" in self.kiosk:
+            r.DSLV3 = s.WLV * k.RF_FROST
+        else:
+            r.DSLV3 = torch.tensor(0., dtype=DTYPE)
+
+        # leaf death equals maximum of water stress, shading and frost
+        r.DSLV = torch.max(torch.stack([r.DSLV1, r.DSLV2, r.DSLV3]))
+
+        # Determine how much leaf biomass classes have to die in states.LV,
+        # given the a life span > SPAN, these classes will be accumulated
+        # in DALV.
+        # Note that the actual leaf death is imposed on the array LV during the
+        # state integration step.
+        DALV = torch.tensor(0., dtype=DTYPE)
+        if p.SPAN.requires_grad:  # replacing hard threshold `if lvage > p.SPAN``
+            sharpness = 1000.0  # FIXEME
+            for lv, lvage in zip(s.LV, s.LVAGE, strict=False):
+                weight = torch.sigmoid((lvage - p.SPAN) * sharpness)
+                DALV = DALV + weight * lv
+        else:
+            for lv, lvage in zip(s.LV, s.LVAGE, strict=False):
+                if lvage > p.SPAN:
+                    DALV = DALV + lv
```

Comment:
I'm not sure about this approach. The code's behavior is now dependent on whether or not grad is required for SPAN. That isn't very transparent, but it could fit the current situation of the project. Mut maybe we should discuss this.
An alternative would be to have 2 separate implementations for either case. This would cause code duplication of course which we could solve as well, but that could get a little convoluted as well.